Fix some Clippy nags

oherrala opened the pull request:
This updates minimum Rust version to 1.13.0 because of `impl From<u8> for char`.

BurntSushi commented:
Thanks for this! Unfortunately, I have some concerns:

* Bumping the minimum Rust version to satisfy Clippy is not something I agree with. The juice ain't worth the squeeze.
* I don't agree with a number of fixes in this PR. Some of them I do agree with, particularly around the use of `is_empty` instead of `len() == 0`.

If you want to reduce this PR down to just the `is_empty` fixes, then that would be good!

oherrala commented:
> Bumping the minimum Rust version to satisfy Clippy is not something I agree with. The juice ain't worth the squeeze.

With 1.13 we get the `char::from(_: u8) -> char` which can be used to remove the call to `char::from_u32(_: u32) -> Option<char>`. This change removes one `unwrap()` from the code (right now it should never panic), but using `char::from(u8)` makes the code safe of panics.

> If you want to reduce this PR down to just the is_empty fixes, then that would be good!

I can do that. Thanks!

BurntSushi commented:
> This change removes one unwrap() from the code (right now it should never panic), but using char::from(u8) makes the code safe of panics.

Yes, I understand why you did it. What I'm saying is that it's not worth it. I am simply never going to bump the minimum Rust version of a crate for something like that. The code is *already* safe from panics.

With all that said, we don't need `char::from(u8)` here. A cast will do. I suspect `b as char` will satisfy both our desires (fewer unwraps and no bumping of the minimum Rust version).

oherrala commented:
Reworked the pull request with only the `.is_empty()` fixes and the on cast from u8 to char.

What should we do with the (now removed) `.fold(0, |a, b| a + b)` -> `.sum::<usize>()` changes?

BurntSushi commented:
Oh, sorry. As long as sum is available on Rust 1.12, then that change should be fine too.

oherrala commented:
Apparently `Iterator::sum()` was introduced in 1.11.0. Added those fixes also.

oherrala commented on src/full.rs:
```diff
@@ -44,23 +44,23 @@ impl<P: AsRef<[u8]>> FullAcAutomaton<P> {
         self.pats.iter()
             .map(|p| vec_bytes() + p.as_ref().len())
-            .fold(0, |a, b| a + b)
+            .sum::<usize>()
```
Yes, I'm not sure why it's not inferred correctly:

```
error[E0283]: type annotations required: cannot resolve `_: std::iter::Sum<usize>`
   --> src/lib.rs:244:14
    |
244 |             .sum()
    |              ^^^
```

BurntSushi replied:
Weird, ah well. Thanks!

BurntSushi commented:
@oherrala Thanks again for this PR! And thanks for your patience with my perhaps idiosyncratic requirements. :-)
